dsa/intermediate/Hash-table.md

```diff
@@ -0,0 +1,131 @@
+# Hash Table
+
+A hash table (or hash map) is a fundamental data structure in computer science, used for fast data retrieval. Here’s an in-depth look at hash tables:
+
+## Overview
+A hash table is a collection of key-value pairs, where each key is mapped to a value. The key is transformed into an index in an array through a process called hashing. This index is used to quickly locate the value associated with the key.
+
+## Key Concepts
+1. *Hash Function:* This function takes an input (the key) and returns an integer (the hash code), which is used as an index to store the corresponding value in the array. A good hash function distributes keys uniformly across the array.
+2. *Buckets:* Each index in the array is referred to as a bucket. In cases where multiple keys hash to the same index, these keys are stored in the same bucket.
+3. *Collision Handling:* Collisions occur when multiple keys hash to the same index. Common collision handling techniques include:
+      - Chaining: Each bucket points to a linked list of entries that map to the same index.
+      - Open Addressing: When a collision occurs, the algorithm searches for the next open bucket using a probing sequence (linear probing, quadratic probing, or double hashing).
+
+## Operations
+**Insertion:**
+
+Compute the hash code of the key using the hash function.
+Map the hash code to an index in the array.
+Store the key-value pair in the corresponding bucket. Handle collisions appropriately.
+
+**Search:**
+
+Compute the hash code of the key.
+Map the hash code to an index.
+Check the corresponding bucket. If using chaining, search the linked list. If using open addressing, follow the probing sequence until the key is found or an empty bucket is reached.
+
+**Deletion:**
+
+Compute the hash code of the key.
+Map the hash code to an index.
+Locate the key in the bucket and remove the key-value pair. Adjust the structure to maintain efficiency (e.g., rehash elements if necessary).
+
+## Complexity
+Average Time Complexity: O(1) for insert, delete, and search operations in a well-implemented hash table with a good hash function and a load factor that avoids excessive collisions.
+
+Worst-Case Time Complexity: O(n) for insert, delete, and search operations if all keys hash to the same index (highly unlikely with a good hash function).
+
+## Load Factor
+The load factor (α) is the ratio of the number of entries to the number of buckets in the hash table. A high load factor can lead to more collisions, affecting performance. A common practice is to resize the hash table (rehash) when the load factor exceeds a certain threshold.
+
+
+​Load Factor = No. of entries /  No. of Buckets
```

Comment:
Replace `​Load Factor = No. of entries / No. of Buckets` to `$$​Load Factor = No. of entries / No. of Buckets$$`